tests/repo_finder/test_provenance_extractor.py

```diff
@@ -0,0 +1,474 @@
+    """Test removing Git attestation from Witness V0.1 GitHub provenance."""
+    attestations = json_extract(witness_github_provenance, ["predicate", "attestations"], list)
+    _json_modify(witness_github_provenance, ["predicate", "attestations"], attestations[:1])
+    with pytest.raises(ProvenanceExtractionException):
+        _perform_provenance_comparison(witness_github_provenance, "", "")
+
+
+@pytest.mark.parametrize(
+    ("type_", "predicate_type"),
+    [
+        ("https://in-toto.io/Statement/v0.1", "https://slsa.dev/provenance/v1"),
+        ("https://in-toto.io/Statement/v1", "https://slsa.dev/provenance/v0.2"),
+        ("https://in-toto.io/Statement/v1", "https://slsa.dev/provenance/v0.1"),
+        ("https://in-toto.io/Statement/v1", "https://witness.testifysec.com/attestation-collection/v0.1"),
+    ],
+)
+def test_invalid_type_payloads(type_: str, predicate_type: str) -> None:
+    """Test payloads with invalid type combinations."""
+    payload_text = '{ "_type": ' + f'"{type_}",' + ' "predicateType": ' + f'"{predicate_type}",'
+    payload_text = f"{payload_text}" + '"subject": [], "predicate": {} }'
+    payload = json.loads(payload_text)
+    with pytest.raises(ProvenanceExtractionException):
+        _perform_provenance_comparison(payload, "", "")
+
+
+def _perform_provenance_comparison(payload: dict[str, JsonType], expected_repo: str, expected_commit: str) -> None:
+    """Accept a provenance and extraction function, assert the extracted values match the expected ones."""
+    provenance = validate_intoto_payload(payload)
+    repo, commit = extract_repo_and_commit_from_provenance(provenance)
+    assert expected_repo == repo
+    assert expected_commit == commit
+
+
+def _json_modify(entry: dict[str, JsonType], keys: list[str], new_value: JsonType) -> None:
+    """Modify the value found by following the list of depth-sequential keys inside the passed JSON dictionary.
+
+    The found value will be overwritten by the new_value parameter.
+    If new_value is None, the value will be removed.
+    If the final key does not exist, it will be created as new_value.
+    """
+    target = entry
+    for index, key in enumerate(keys):
+        if key not in target:
+            if index == len(keys) - 1:
+                # Add key.
+                target[key] = new_value
+                return
+            raise JsonExtractionException(f"JSON key not found: {key}")
+        next_target = target[key]
+        if index == len(keys) - 1:
+            if new_value is None:
+                # Remove value.
+                del target[key]
+            else:
+                # Replace value
+                target[key] = new_value
+        else:
+            if not isinstance(next_target, dict):
+                raise JsonExtractionException(f"Cannot extract value from non-dict type: {str(type(next_target))}")
+            target = next_target
```

Comment:
Nit: This may not really matter because this is test code anyway. However, I would have the same comment as the one I had with the `json_extract` function. I feel like the `for` loop is trying to do too many things which results in unnecessary branching within it. It can definitely be separated into a two-step process: first traversing the dict, then modifying the value.

Also, I personally think for test code, we do not need to be super strict about throwing and catching exceptions. If anything, I would personally go against trying to error-handling test code like it's production code. Errors in test code mean the test is wrongly set up, and not trying to swallow them or converting them to some custom error type is a good thing as it helps us catch our own mistakes.